src/App/AutoTransaction.h

```diff
@@ -31,7 +31,9 @@ class Application;
 class AppExport AutoTransaction {
 private:
     /// Private new operator to prevent heap allocation
-    void* operator new(size_t size);
+    void* operator new(size_t size) {
+        return (void*)size;
```

Comment:
I know it's academic since the point is that this can't be called, but it seems that just returning `nullptr` would be more idiomatic.

Reply:
nullptr is also what MSVC proposed but here: https://stackoverflow.com/questions/5909636/overloading-operator-new

someone wrote "The operator new should return a buffer of the size specified as the first argument."

Reply:
Right -- if we were actually implementing this operator, it would return a buffer whose size is equal to the size passed in. Again, this is totally academic, the point of the code is that it can't be run, and trying to call it will give a compiler error about a private operator new. But it's a little jarring as a reviewer to see the location of size being returned.

Reply:
Fine with me of course. I changed it now to a nullptr. If this is now fine, please squash-merge.

Reply:
I don't think you understood the essence of the link I posted. Making the new operator private only serves the purpose to disallow in client code to create an object on the heap and thus it suffices to **declare** it but a **definition** is not needed.

The error you got was because of changing the order of includes and thus the compiler got AutoTransaction.h as first that doesn't include a header file the pulls in the header where "size_t" is defined (because it's only a typedef of the actual type).

In the first answer of the SO question exactly this was said.

Now this PR makes things even worse because it creates hundreds of warnings:
```
AutoTransaction.h:35:9: warning: 'operator new' should not return a null pointer unless it is declared 'throw()' or 'noexcept' [-Wnew-returns-null]
AutoTransaction.h:34:31: warning: unused parameter 'size' [-Wunused-parameter]
AutoTransaction.h:126:9: warning: 'operator new' should not return a null pointer unless it is declared 'throw()' or 'noexcept' [-Wnew-returns-null]
AutoTransaction.h:125:31: warning: unused parameter 'size' [-Wunused-parameter]
```

Reply:
> Making the new operator private only serves the purpose to disallow in client code to create an object on the heap and thus it suffices to declare it but a definition is not needed

I understood the first part. But when no definition is needed, why does MSVC want one? Maybe I trust MSVC too much and MSVC proposed exactly what I merged.

> The error you got was because of changing the order of includes and thus the compiler got AutoTransaction.h as first that doesn't include a header file the pulls in the header where "size_t" is defined

Yes, that's why I realized I have to have Application.h as header before AutoTransaction.h. This part is already fixed. However, since MSVC kept saying me that it wants a definition block, I made this PR.

> Now this PR makes things even worse because it creates hundreds of warnings:

Sh... I don't get them. Do they disappear when you return `(void*)size`? If not, I will revert the change.

Reply:
I see them in the CI and reverted the PR.

Reply:
> But when no definition is needed, why does MSVC want one?

What does it say exactly? And does it say for every cpp file that includes AutoTransaction.h directly or indirectly?

> Maybe I trust MSVC too much and MSVC proposed exactly what I merged.

Maybe yes. With newer versions it's not as bad as with older versions that reported many false-positives that we had to suppress with pragmas.

> Yes, that's why I realized I have to have Application.h as header before AutoTransaction.h. This part is already fixed.

And this is not needed because AutoTransation.h requires cstddef instead for the type size_t

Reply:
The warning only appear in AutoTransaction.h:
![ReItjRrLHL](https://user-images.githubusercontent.com/1828501/156948408-45a5e6ed-6c48-459d-96dd-2d3df074dea4.png)

> And this is not needed because AutoTransation.h requires cstddef instead for the type size_t

Since the operator is defined within class Application I thought, Application.h is nevertheless necessary. However, I know my limitations of never learned C++, only learning by doing. Therefore the best is that I don't act here further.

---------------
Concerning my work with the headers. I am ready for FC 0.20 and also won't have tome to look at the yet unreviewed SoF* files in GUI the next days. I have 2 pending PRs because I am not sure if the CI compiler will bring errors or warnings.

Reply:
> Since the operator is defined within class Application I thought, Application.h is nevertheless necessary.

No, it's not. In C++ it's possible to implement a new allocator for each class individually. And that you can see that the function is declared inside the class declaration.

Well, to me this MSVC hint is a false-positive because the code checker apparently doesn't realize the purpose of it. I wonder what the code checker says when you e.g. open the file Writer.h. Does it also complain about the lack of the implementation of the copy constructor and assignment operator of the class Writer?

Reply:
> Does it also complain about the lack of the implementation of the copy constructor and assignment operator of the class Writer?

No, it accepts all:
![devenv_ZAYhS0akww](https://user-images.githubusercontent.com/1828501/157017897-cfd27374-dc80-4b6e-80ca-beb8a61a2fdd.png)

Reply:
In order to make the disallowed use of the new function more explicit you could write:
`void* operator new (std::size_t) = delete;`

Reply:
> In order to make the disallowed use of the new function more explicit you could write: `void* operator new (std::size_t) = delete;`

This works. here is the PR: #6526